pkg/integrations/v2/eventhandler/eventhandler.go

```diff
@@ -0,0 +1,454 @@
+	// this won't increase cardinality but maybe improves UX?
+	labels[model.LabelName("name")] = model.LabelValue(obj.Name)
+	labels[model.LabelName("namespace")] = model.LabelValue(obj.Namespace)
+
+	// we add these fields to the log line to reduce label bloat and cardinality
+	if event.Action != "" {
+		msg.WriteString(fmt.Sprintf("action=%s ", event.Type))
+	}
+	if obj.APIVersion != "" {
+		msg.WriteString(fmt.Sprintf("apiversion=%s ", obj.APIVersion))
+	}
+	if obj.ResourceVersion != "" {
+		msg.WriteString(fmt.Sprintf("resourceversion=%s ", obj.ResourceVersion))
+	}
+
+	// useful for debugging, can omit when shipping this
+	if event.ResourceVersion != "" {
+		msg.WriteString(fmt.Sprintf("evRV=%s ", event.ResourceVersion))
+	}
+
+	if event.ReportingInstance != "" {
+		msg.WriteString(fmt.Sprintf("reportinginstance=%s ", event.ReportingInstance))
+	}
+	if event.ReportingController != "" {
+		msg.WriteString(fmt.Sprintf("reportingcontroller=%s ", event.ReportingController))
+	}
+	if event.Source.Component != "" {
+		msg.WriteString(fmt.Sprintf("sourcecomponent=%s ", event.Source.Component))
+	}
+	if event.Source.Host != "" {
+		msg.WriteString(fmt.Sprintf("sourcehost=%s ", event.Source.Host))
+	}
+	if event.Reason != "" {
+		msg.WriteString(fmt.Sprintf("reason=%s ", event.Reason))
+	}
+	if event.Type != "" {
+		msg.WriteString(fmt.Sprintf("type=%s ", event.Type))
+	}
+	if event.Count != 0 {
+		msg.WriteString(fmt.Sprintf("count=%d ", event.Count))
+	}
+
+	msg.WriteString(fmt.Sprintf("msg=%q", event.Message))
+
+	return labels, msg.String(), nil
+}
+
+func getTimestamp(event *v1.Event) time.Time {
+	if !event.LastTimestamp.IsZero() {
+		return event.LastTimestamp.Time
+	}
+	return event.EventTime.Time
+}
+
+func newEntry(msg string, ts time.Time, labels model.LabelSet) api.Entry {
+	entry := logproto.Entry{Timestamp: ts, Line: msg}
+	return api.Entry{Labels: labels, Entry: entry}
+}
+
+// maintain "last event" state and write it out to disk as necessary
+func (eh *EventHandler) updateLastEvent(e *v1.Event, eventTs time.Time) error {
+	// resource handler fns don't run concurrently so we probably don't need this...
+	eh.Lock()
+	defer eh.Unlock()
+
+	eventRv := e.ResourceVersion
+
+	if eh.LastEvent == nil {
+		// startup
+		eh.LastEvent = &ShippedEvents{Timestamp: eventTs, RvMap: make(map[string]struct{})}
+		eh.LastEvent.RvMap[eventRv] = struct{}{}
+		return nil
+	}
+
+	// if timestamp is the same, add to map
+	if eh.LastEvent != nil && eventTs.Equal(eh.LastEvent.Timestamp) {
+		eh.LastEvent.RvMap[eventRv] = struct{}{}
+		return nil
+	}
+
+	// if timestamp is different, write out the current last event and create
+	// a new lastEvent struct
+	err := eh.writeOutLastEvent()
+	if err != nil {
+		return err
+	}
+	eh.LastEvent = &ShippedEvents{Timestamp: eventTs, RvMap: make(map[string]struct{})}
+	eh.LastEvent.RvMap[eventRv] = struct{}{}
+	return nil
+}
+
+func (eh *EventHandler) writeOutLastEvent() error {
```

Comment:
Few concerns here:

1. Does this need to be appending to the end of the file every time? It seems like we never read intermediate events. 
2. This might have high IOPS on busy clusters. It's probably better to flush to the file in the background every ~10 seconds or so and overwrite its entire contents.

Reply:
It def does not need to append every time - we only need the last event for restore purposes

I prefer your approach over what I'm doing now... was originally gonna truncate every 10s or so but would rather just cut down IOPS and flush every interval. Will make this change.